File: Dynamic Programming/1143. Longest Common Subsequence.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
var longestCommonSubsequence = function(text1, text2) {
    const m = text1.length;
    const n = text2.length;
    const dp = new Array(m + 1).fill(0).map(i => new Array(n + 1).fill(0));
    for(let i = 1; i <= m; i++) {
        for(let j = 1; j <= n; j++) {
            if (text1[i - 1] === text2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1]; 
            } else {
                dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }
    return dp[i][j];
}

var longestCommonSubsequence = function(text1, text2) { // Based on the state transform equation, we can reduce the memory usage by using a rolling array
  const m = text1.length;
  const n = text2.length;
  let dp = new Array(n + 1).fill(0);
  for(let i = 1; i <= m; i++) {
      const temp = new Array(n + 1).fill(0);
      for(let j = 1; j <= n; j++) {
          if (text1[i] === text2[j]) {
              temp[j] = 1 + dp[j - 1];
          } else {
              temp[j] = Math.max(temp[j - 1], dp[j]); 
          }
      }
      dp = temp;
  }
  return dp[n];
}

File: Golang/DynamicProgramming/1143.longest-common-subsequence.go
package main

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func longestCommonSubsequenceDP2D(text1 string, text2 string) int {
	ret, length1, length2 := 0, len(text1), len(text2)
	dp := make([][]int, length1+1)
	for i := 0; i <= length1; i++ {
		dp[i] = make([]int, length2+1)
	}
	for i := 1; i <= length1; i++ {
		for j := 1; j <= length2; j++ {
			if text1[i-1] == text2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
			ret = max(ret, dp[i][j])
		}
	}
	return ret
}

func longestCommonSubsequenceDP1D(text1 string, text2 string) int {
	ret, length1, length2 := 0, len(text1), len(text2)
	dp := make([]int, length2+1)
	for i := 1; i <= length1; i++ {
		temp := make([]int, length2+1)
		for j := 1; j <= length2; j++ {
			if text1[i-1] == text2[j-1] {
				temp[j] = dp[j-1] + 1
			} else {
				temp[j] = max(dp[j], temp[j-1])
			}
			ret = max(ret, temp[j])
		}
		dp = temp
	}
	return ret
}
